rust/src/countme/cookie.rs

```diff
@@ -0,0 +1,183 @@
+}
+fn parse_cookie_v0(line: &str, now: i64) -> Option<Cookie> {
+    let mut split= line.split(' ');
+    // Ignore version
+    match split.next() {
+        None => return None,
+        Some(_s) => (),
+    }
+    let epoch = match split.next() {
+        None => return None,
+        Some(s) => {
+            match s.parse::<i64>() {
+                Err(_e) => return None,
+                Ok(i) => i,
+            }
+        },
+    };
+    let window = match split.next() {
+        None => return None,
+        Some(s) => {
+            match s.parse::<i64>() {
+                Err(_e) => return None,
+                Ok(i) => i,
+            }
+        },
+    };
+    Some(Cookie {
+        epoch,
+        window,
+        now,
+    })
+}
+
+/// Load cookie timestamps from persistent directory if it exists.
+/// Returns a default cookie (counting never started) in all other cases.
+pub fn new() -> Cookie {
+    let now = Utc::now().timestamp();
+    // Start default window at COUNTME_OFFSET to avoid negative values
+    let c = Cookie {
+        epoch: now,
+        window: COUNTME_OFFSET,
+        now,
+    };
+    let cookie: path::PathBuf = [STATE_DIR, COUNTME_COOKIE].iter().collect();
+    if !cookie.exists() {
+        return c;
+    }
+    let file = match fs::File::open(&cookie) {
+        Err(e) => {
+            println!("Could not open '{}': {}", cookie.display(), e);
+            return c;
+        },
+        Ok(f) => f,
+    };
+    let mut reader = BufReader::new(file);
+    let mut line = String::new();
+    // We only read the first line and ignore everything else from the file
+    match reader.read_line(&mut line) {
+        Err(e) => {
+            println!("Could not read from '{}': {}", cookie.display(), e);
+            return c;
+        },
+        Ok(_c) => (),
+    };
+    let version = match line.split(' ').next() {
+        None => {
+            println!("Could not read cookie version from '{}'", cookie.display());
+            return c;
+        },
+        Some(s) => s,
+    };
+    match version {
+        "0" => {
+            match parse_cookie_v0(&line, now) {
+                None => {
+                    println!("Could not parse as version 0 from '{}'", cookie.display());
+                    c
+                }
+                Some(c) => c,
+            }
+        },
+        _ => {
+            println!("Unknown version '{}'", cookie.display());
+            c
+        },
+    }
+}
+
+impl Cookie {
+    fn epoch_window(&self) -> i64 {
+        (self.epoch - COUNTME_OFFSET) / COUNTME_WINDOW
+    }
+
+    fn previous_window(&self) -> i64 {
+        (self.window - COUNTME_OFFSET) / COUNTME_WINDOW
+    }
+
+    fn current_window(&self) -> i64 {
+        (self.now - COUNTME_OFFSET) / COUNTME_WINDOW
+    }
+
+    /// Are we still in the same window as the one we loaded from the cookie?
+    pub fn existing_window(&self) -> bool {
+        self.current_window() <= self.previous_window()
+    }
```

Comment:
Minor: minimal unit tests for these would be nice (it also help other devs understand how it works!)